src/main/java/com/feedzai/commons/sql/abstraction/batch/AbstractBatch.java

```diff
@@ -281,40 +324,71 @@ public void flush() {
             bufferLock.unlock();
         }
 
+        long start = System.currentTimeMillis();
         try {
             flushTransactionLock.lock();
-            final long start = System.currentTimeMillis();
+            start = System.currentTimeMillis();
 
             // begin the transaction before the addBatch calls in order to force the retry
             // of the connection if the same was lost during or since the last batch. Otherwise
             // the addBatch call that uses a prepared statement will fail
             de.beginTransaction();
 
-            for (BatchEntry entry : temp) {
+            for (final BatchEntry entry : temp) {
                 de.addBatch(entry.getTableName(), entry.getEntityEntry());
             }
 
             de.flush();
             de.commit();
             logger.trace("[{}] Batch flushed. Took {} ms, {} rows.", name, (System.currentTimeMillis() - start), temp.size());
-        } catch (Exception e) {
-            logger.error(dev, "[{}] Error occurred while flushing.", name, e);
-            /*
-             * We cannot try any recovery here because we don't know why it failed. If it failed by a Constraint
-             * violation for instance, we cannot try it again and again because it will always fail.
-             *
-             * The idea here is to hand the entries that could not be added and continue right where we were.
-             *
-             * So, temp will only have the entries until it failed, this can be all the entries or until de.addBatch() failed;
-             * and buffer will have the events right after the failure.
-             */
-            onFlushFailure(temp.toArray(new BatchEntry[temp.size()]));
+        } catch (final Exception e) {
+            if (this.maxFlushRetries > 0) {
+                logger.warn(dev, "[{}] Error occurred while flushing. Retrying.", name, e);
+            }
+
+            boolean success = false;
+            int retryCount;
+
+            for (retryCount = 0; retryCount < this.maxFlushRetries && !success; retryCount++) {
+                try {
+                    Thread.sleep(this.flushRetryDelay);
+
+                    // If the connection was established, we might need a rollback.
+                    if (de.checkConnection() && de.isTransactionActive()) {
+                        de.rollback();
+                    }
+
+                    // begin the transaction before the addBatch calls in order to force the retry
+                    // of the connection if the same was lost during or since the last batch. Otherwise
+                    // the addBatch call that uses a prepared statement will fail
+                    de.beginTransaction();
+
```

Comment:
We are repeating the same code the first time we are trying to execute the batches and on the repeat. We should extract that (also the method is already quite large).